fence/agents/powerman/fence_powerman.py

```diff
@@ -0,0 +1,300 @@
+#!/usr/bin/env python
+import os
+import time
+from datetime import datetime
+import sys
+import subprocess
+import re
+import atexit
+sys.path.append("@FENCEAGENTSLIBDIR@")
+from fencing import *
+from fencing import run_delay
+import logging
+
+
+#BEGIN_VERSION_GENERATION
+RELEASE_VERSION="Powerman Fencing Agent"
+REDHAT_COPYRIGHT=""
+BUILD_DATE=""
+#END_VERSION_GENERATION
+
+
+#### important!!! #######
+class PowerMan:
+	"""Python wrapper for calling powerman commands
+
+	This class makes calls to a powerman deamon for a cluster of computers.
+	The make-up of such a call looks something like:
+		$ pm -h elssd1:10101 <option> <node>
+	where option is something like --off, --on, --cycle and where node is 
+	elssd8, or whatever values are setup in powerman.conf (***this is key, 
+	because otherwise this code will not work!)
+	"""
+	program_name = "powerman"
+
+	def __init__(self, server_name, port="10101"):
+		"""
+		Args:
+			server_name: (string) host or ip of powerman server
+			port: (str) port number that the powerman server is listening on
+		"""
+		self.server_name = server_name
+		self.port = port
+		self.server_and_port = server_name + ":" + str(port)
+		# this is a list of the command and its options. For example:
+		# ['powerman', '--server-host', 'elssd1:10101']
+		self.base_cmd = [
+			self.program_name, 
+			"--server-host", 
+			self.server_and_port
+		]
+
+	def _run(self, cmd, only_first_line):
+		# Args:
+		#   cmd: (list) commands and arguments to pass to the program_name
+
+		# add the 2 command lists together to get whole command to run
+		run_this = self.base_cmd + cmd 
+		try:
+			popen = subprocess.Popen(run_this, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
+			out = popen.communicate()
+		except OSError as e:
+			logging.error("_run command error: %s\n", e)
+			sys.exit(1)
+		if only_first_line == True:
+			result_line = out[0].decode().strip()
+			return (result_line, popen.returncode)
+		else:
+			result_list = []
+			for line in out:
+				result_list.append(line)
+			return (result_list, popen.returncode)
+
+	def is_running(self):
+		"""simple query to see if powerman server is responding. Returns boolean"""
+		cmd = ["-q"] # just check if we get a response from the server
+		result, ret_code = self._run(cmd, True)
+		if ret_code != 0:
+			return False
+		return True
+
+	def on(self, host):
+		logging.debug("PowerMan on: %s\n", host)
+		cmd = ["--on", host]
+		try:
+			result, ret_code = self._run(cmd, True)
+		except OSError as e:
+			logging.error("PowerMan Error: The command '--on' failed: %s\n", e)
+			return -1
+		except ValueError as e:
+			logging.error("PowerMan Error: Popen: invalid arguments: %s\n", e)
+			return -1
+		logging.debug("pm.on result: %s ret_code: %s\n", result, ret_code)
+
+		## For some devices, the "on" or "off" command can fail silently,
```

Comment:
yes, to be precise we don't care about return codes from set functions at all. Because quite often device really believes that everything is okay but it is not.